tests/validators.py

```diff
@@ -0,0 +1,85 @@
+from unittest import TestCase
+
+from adapters.validators import (
+    EMail, Length, Schema, Type, UnicodeNormalize, ValidationError,
+)
+
+
+class ValidatorsTest(TestCase):
+    def setUp(self):
+        class UsernameSchema(Schema):
+            inputs = {'username'}
+            outputs = {'username', 'username_normalized'}
+            validators = [
+                Type('username', str),
+                Length('username', min=4, max=100),
+                UnicodeNormalize('username', out='username_normalized'),
+            ]
+
+        class EMailSchema(Schema):
+            inputs = {'email'}
+            validators = [
+                EMail('email'),
+            ]
+
+        class RegistrationSchema(Schema):
+            depends = [UsernameSchema, EMailSchema]
+
+            def username_in_email(self):
+                if self.username_normalized not in self.email:
```

Comment:
I also thought about explicitly specifying which fields the validator expects. But then I think it gets pretty hard to specify just a single validator if you have to explicitly set all the fields.

Maybe having a method doesn't express how I meant the validator functions to be called well enough. I thought about them having a signature like `def validator(data)` and thought that maybe a method on a `Schema` could be a special case where `self` is equivalent to `data`. On second thought I guess

```python
@staticmethod
def username_in_mail(data):
   ...
```
is clearer.